File: sample-app/src/main/java/org/granchi/whatnow/framework/dagger/WhatNowModule.java
package org.granchi.whatnow.framework.dagger;

import android.app.Application;

import org.granchi.hollywood.Actor;
import org.granchi.hollywood.CompositeModel;
import org.granchi.hollywood.HollywoodApplication;
import org.granchi.hollywood.Model;
import org.granchi.hollywood.ModelExceptionHandler;
import org.granchi.hollywood.ReinstateModelExceptionHandler;
import org.granchi.whatnow.StubActor;
import org.granchi.whatnow.TasksModel;
import org.granchi.whatnow.WhatNowApplication;
import org.granchi.whatnow.WhatNowPreferencesModel;

import java.util.Arrays;
import java.util.Collection;

import dagger.Module;
import dagger.Provides;

/**
 * Dagger module for the app.
 *
 * @author serandel
 */
@Module
public class WhatNowModule {
    private final WhatNowApplication app;

    public WhatNowModule(WhatNowApplication app) {
        this.app = app;
    }

    @Provides
    WhatNowApplication provideWhatNowApplication() {
        return app;
    }

    @Provides
    Application provideApplication() {
        return app;
    }

    @Provides
    Model provideInitialModel(TasksModel tasksModel,
                              WhatNowPreferencesModel preferencesModel) {
        return new CompositeModel(tasksModel, preferencesModel);
    }

    @Provides
    HollywoodApplication provideHollywoodApp(Model initialModel,
                                             Collection<Actor> actors,
                                             ModelExceptionHandler exceptionHandler) {
        return new HollywoodApplication(initialModel, actors, exceptionHandler);
    }

//
//    @Provides
//    Crew.Factory<SingleInstanceActorMetadata> provideCrewFactory() {
//        return models -> new SingleInstanceCrew(metadata -> {
//            try {
//                return (Actor<SingleInstanceActorMetadata>) Class.forName(metadata
// .getActorClass())
//                                                                 .getConstructor()
//                                                                 .newInstance();
//            } catch (Exception e) {
//                throw new RuntimeException(e);
//            }
//        }, models);
//    }
//
//    @Provides
//    ModelExceptionHandler provideExceptionHandler(
//            WhatNowApplication app) {
//        return (model, action, exception) -> {
//            app.showError("Exception");
//            return model;
//        };
//    }

    @Provides
    Collection<Actor> provideActors() {
        return Arrays.asList(new StubActor());
    }

    @Provides
    ModelExceptionHandler provideModelExceptionHandler() {
        return new ReinstateModelExceptionHandler();
    }
}

File: sample-app/src/main/kotlin/org/granchi/whatnow/framework/dagger/WhatNowModule.kt
package org.granchi.whatnow.framework.dagger

import android.app.Application
import dagger.Module
import dagger.Provides
import org.granchi.hollywood.*
import org.granchi.whatnow.StubActor
import org.granchi.whatnow.TasksModel
import org.granchi.whatnow.WhatNowApplication
import org.granchi.whatnow.WhatNowPreferencesModel
import java.util.*

/**
 * Dagger module for the app.
 *
 * It's open to let a debug module inherit from it.
 */
@Module
open class WhatNowModule(private val app: WhatNowApplication) {
    @Provides
    fun provideWhatNowApplication(): WhatNowApplication {
        return app
    }

    @Provides
    fun provideApplication(): Application {
        return app
    }

    @Provides
    @JvmSuppressWildcards
            // Needed so the stubs aren't generated having a Collection<? extends Actor>, for some reason
    fun provideHollywoodApp(initialModel: Model,
                            actors: Collection<Actor>,
                            exceptionHandler: ModelExceptionHandler): HollywoodApplication {
        return HollywoodApplication(initialModel,
                actors,
                exceptionHandler)
    }

    @Provides
    fun provideInitialModel(tasksModel: TasksModel,
                            preferencesModel: WhatNowPreferencesModel): Model {
        return CompositeModel(tasksModel, preferencesModel)
    }

    @Provides
    fun provideActors(): Collection<Actor> {
        return Arrays.asList<Actor>(StubActor())
    }

    @Provides
    fun provideModelExceptionHandler(): ModelExceptionHandler {
        return ReinstateModelExceptionHandler()
    }
}
